src/Microsoft.VisualStudio.Composition/Configuration/Formatters/RuntimePartFormatter.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Composition.Formatter
+{
+    using System.Collections.Immutable;
+    using MessagePack;
+    using MessagePack.Formatters;
+    using Microsoft.VisualStudio.Composition.Reflection;
+    using static Microsoft.VisualStudio.Composition.RuntimeComposition;
+
+#pragma warning disable CS3001 // Argument type is not CLS-compliant
+
+    public class RuntimePartFormatter : IMessagePackFormatter<RuntimePart>
+    {
+        /// <inheritdoc/>
+        public RuntimePart Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
+        {
+            var importingCtor = default(MethodRef);
+            IReadOnlyList<RuntimeComposition.RuntimeImport> importingCtorArguments = ImmutableList<RuntimeComposition.RuntimeImport>.Empty;
+
+            TypeRef typeRef = options.Resolver.GetFormatterWithVerify<TypeRef>().Deserialize(ref reader, options);
+            IReadOnlyList<RuntimeExport> exports = MessagePackCollectionFormatter<RuntimeExport>.DeserializeCollection(ref reader, options);
+            bool hasCtor = options.Resolver.GetFormatterWithVerify<bool>().Deserialize(ref reader, options);
+
+            if (hasCtor)
+            {
+                importingCtor = options.Resolver.GetFormatterWithVerify<MethodRef?>().Deserialize(ref reader, options);
+                importingCtorArguments = MessagePackCollectionFormatter<RuntimeImport>.DeserializeCollection(ref reader, options);
+            }
+
+            IReadOnlyList<RuntimeImport> importingMembers = MessagePackCollectionFormatter<RuntimeImport>.DeserializeCollection(ref reader, options);
+            IReadOnlyList<MethodRef> onImportsSatisfiedMethods = MessagePackCollectionFormatter<MethodRef>.DeserializeCollection(ref reader, options);
+
+            string sharingBoundary = options.Resolver.GetFormatterWithVerify<string>().Deserialize(ref reader, options);
+
+            return new RuntimeComposition.RuntimePart(
+                      typeRef,
+                      importingCtor,
+                      importingCtorArguments,
+                      importingMembers,
+                      exports!,
+                      onImportsSatisfiedMethods,
+                      sharingBoundary);
+        }
+
+        /// <inheritdoc/>
+        public void Serialize(ref MessagePackWriter writer, RuntimePart value, MessagePackSerializerOptions options)
+        {
+            options.Resolver.GetFormatterWithVerify<TypeRef>().Serialize(ref writer, value.TypeRef, options);
```

Comment:
I am wondering if a custom formatter is really needed here since MessagePack should be able to serialize ImportingConstrcutorOrFactoryMethodRef as null. Otherwise it looks like you are just using existing formatters.